docs/console-access.md

```diff
@@ -0,0 +1,340 @@
+
+Choice of access control methods. Either QEMU can mandate client certificates
+and perform access control checks on the certificate identity (planned, but not
+yet supported in QEMU & libvirt), or protocol specific authentication must be
+used.
+
+Only supports VNC and SPICE, until QEMU is able to do fine grained ACL checks
+against client certs.
+
+
+### 2. HTTP CONNECT proxy
+
+QEMU on the compute node running a TCP socket with TLS and x509 certificates
+enabled.
+
+Compute node certs must be issued by a CA that is trusted by all tenant
+users and new certs acquired for each new compute node.
+
+One or more proxy servers able to multiplex to an arbitrary number of backend
+compute nodes
+
+Tenant users make HTTP CONNECT request to proxy server, giving the IP address
+and port on the cluster network that QEMU is listening on
+
+Tenant users directly validate the received server certificates against
+the requested IP + port of the QEMU service.
+
+Choice of access control methods. Either QEMU can mandate client certificates
+and perform access control checks on the certificate identity (planned, but not
+yet supported in QEMU & libvirt), or protocol specific authentication must be
+used.
+
+Only supports VNC and SPICE, until QEMU is able to do fine grained ACL checks
+against client certs.
+
+
+### 3. HTTP websockets proxy
+
+QEMU on the compute node running a TCP socket with TLS and x509 certificates
+enabled.
+
+Compute node certs can be issued by a private CA, which only needs to be
+trusted by the proxy servers. Proxies are issued client certs by the CA
+for mutual auth.
+
+One or more proxy servers able to multiplex to an arbitrary number of backend
+compute nodes
+
+Shared secret between the proxy and tenant user identifies the QEMU service to
+connect to and authenticates the user.
+
+Tenant users validate the received server certificates against the IP + port
+of the proxy server.
+
+The proxy server performs TLS handshake with the QEMU servers and validates its
+certificate.
+
+Protocol specific authentication is not exposed to tenant users.
+
+Supports VNC, SPICE and serial consoles with parity of functionality.
+
+
+Proxy Implementation
+--------------------
+
+For both the HTTP CONNECT and HTTP websockets proxy there needs to be some be some
+intelligence in the proxy. In the HTTP CONNECT proxy, there needs to be code to
+validate that the requested IP+port pair actually corresponds to a valid guest
+console connection. This protects against the proxy being used to attack other
+arbitrary network servers running in the cluster network. In the HTTP websockets
+proxy, there needs to be code to validate the shared secret obtained from the
+client, to resolve it into a IP+port for the QEMU connection. It also needs to
+be prepared to do some intelligent MITM of the connections to handle authentication
+and migration.
+
+The libvirt-console-proxy project aims to provide a general purpose server that
```

Comment:
The design of libvirt-console-proxy is to have separate daemons for doing the proxying, vs resolving the access tokens. The daemon handling proxying is totally generic and does not need any application specific knowledge in it. It has greater security attack surface since it is handling VNC / SPICE protocol logic, and thus it is highly desirable to restrict what it can do. The access token service is application specific, and requires elevated privileges to access the backend mgmt system to deal with the token validation / lookup. Having it in a separate daemon gives it good security isolation from the protocol proxying daemon, as well as allowing easy application specific plugin. Finally, there will likely need to be many more proxying daemons deployed to cope with increasing loads as the number of VMs increases. The token access daemon has minimal runtime overhead so very few are required. 

So overall, the separate daemon design makes much more sense. We would just consume it via a standalone docker image and provide our own token access daemon via an image of our own.